Implement part of #9827: Migrate Google App Engine Mail Functionality to use Mailgun

kevjumba opened the pull request:
## Overview
<!--
READ ME FIRST:
Please answer *both* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

1. This PR fixes or fixes part of #9827 .
2. This PR does the following: 
This PR migrates the google app engine mail api over to using mailgun. 
- The email services send_mail and send_bulk_mail functions are migrated to email_services.py and the platform models.import_email_services() function now only imports the API for posting a query to mailgun. 
- Currently the tests for email functionality on production server are NOT done.

## Essential Checklist

- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The linter/Karma presubmit checks have passed locally on your machine.
- [x] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
  - This lets reviewers restart your CircleCI tests for you.
- [x] The PR is made from a branch that's **not** called "develop".

## PR Pointers

- Oppiabot will notify you when you don't add a PR_CHANGELOG label. If you are unable to do so, please @-mention a code owner (who will be in the Reviewers list), or ask on [Gitter](https://gitter.im/oppia/oppia-chat).
- For what code owners will expect, see the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).
- Make sure your PR follows conventions in the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide), otherwise this will lead to review delays
- Never force push. If you do, your PR will be closed.

kevjumba committed:
Migrate testing in gae_email_services_test from mail_stub

kevjumba committed:
remove accidental commit

kevjumba committed:
Fix linter issues

kevjumba committed:
Remove mail_stub usage from gae_email_services_test and email_subscription_services_test

kevjumba committed:
Change up the format of email test base and migrate email_manager_test over

kevjumba committed:
Fix linting

kevjumba committed:
Merge remote-tracking branch 'upstream/develop' into migration-to-mailgun

kevjumba committed:
Fix linting

seanlip commented on core/tests/test_utils.py:
```diff
@@ -161,6 +164,127 @@ def check_image_png_or_webp(image_string):
 
 
+class EmailMessageMock(object):
+    """Mock for google.appengine.api.email message."""
```
This seems odd. We're getting rid of google.appengine.api.email altogether, so why are we still mocking it?

Instead you should mock mailgun (based on the current API for that) since you want to have a "local stub" for that. Any trace of anything related to google.appengine.api.email should disappear henceforth.

kevjumba replied:
Done

srijanreddy98 commented:
Hi @kevjumba, I have marked your pr as a draft pr because the changes are not yet final. You can check more about that here (https://github.blog/2019-02-14-introducing-draft-pull-requests/). In future prefix your commits with [ci skip] or [skip ci] to prevent CI checks from running. You can learn more about it <a href="https://github.com/oppia/oppia/wiki/Contributing-code-to-Oppia#wip--draft-pull-requests">here</a>.

seanlip commented on core/tests/test_utils.py:
```diff
@@ -161,6 +164,127 @@ def check_image_png_or_webp(image_string):
+
+
+class EmailServicesMock(object):
+    """Mock for google.appengine.api.email."""
```
Same note.

kevjumba replied:
Done

seanlip reviewed (changes requested):
Just a couple of preliminary questions -- the overall idea seems reasonable, but I think you mocked the wrong thing. Also, anything related to google.appengine.api.mail should be deleted completely from the codebase.

Also, did you create the new "sample" label? If so, please don't do that -- don't create new labels, since it messes with the PR workflow. Use existing ones or make your intentions clear in the PR description.

Finally, I don't understand what Python Pillow has to do with this (per the title of your PR). Pillow is an imaging library and that has nothing to do with the Mail API. Please check your PR description carefully.

kevjumba committed:
Use EmailTestBase in gae_email_services_test and email_subscription_test

kevjumba committed:
Migrate reader_test and topic_viewer_test

kevjumba committed:
Add dev mode handling to mailgun api

kevjumba committed:
[skip ci] Migrate classifier_test

seanlip commented:
Hi @kevjumba -- if this PR is not ready for review, could you please manually cancel the running CI jobs (by clicking "Details" in the checks at the bottom)? Thanks!

seanlip commented on core/tests/test_utils.py:
```diff
@@ -161,6 +164,127 @@ def check_image_png_or_webp(image_string):
+
+
+class EmailServicesMock(python_utils.OBJECT):
+    """Mock for mailgun API"""
```
Shouldn't this be a stub for the email service (which has different implementations on dev and prod), rather than the mailgun API?

kevjumba replied:
Done

seanlip commented on core/tests/test_utils.py:
```diff
@@ -161,6 +164,127 @@ def check_image_png_or_webp(image_string):
+        """Reset email dictionary for a new test."""
+        self.emails_dict = {}
+
+    def mock_send_mail(
```
There's lots of non-third-party-specific code here. I think what would be better to do is: limit platform/ to the actual "sending mail" action, and have the code in the email services module be as minimal as possible -- it should only contain the code that interacts with the third-party service. Compose and check everything else in the domain layer.

Also, to make things simpler for yourself, perhaps start by getting rid entirely of gae_email_services.py. You're replicating stuff from that and I think it's making life harder, since doing so is not necessary because the only thing that actually matters is the mailgun implementation.

Finally, this implementation should be self-contained and not refer to email_services at all. You are trying to swap out the one (or at most two) functions in email_services in tests, because the email service in the platforms layer should be doing nothing but making an external call, and you don't want to make external calls in the tests.

I suggest that, before doing anything else on this PR, you create a design doc that lists the functions in the final email_services module and provides the code for their implementation in full. Get sign-off on that before continuing. Then start again. Do the following:

(a) switch the default from gae_email_services to mailgun_email_services, and get rid of gae_email_services altogether. 
(b) modify mailgun_email_services to have the API and code that was approved above. (That code should handle dev/prod mode correctly.)
(c) create stubs for the functions in mailgun_email_services in this file, to be used only in tests.
(d) fix any broken tests.
(e) figure out how to generalize core/platform/models.py to allow a separate email provider implementation to be specified and to have that import the mailgun_email_services file. (It's not really reasonable to consider that a "gae" thing.)

Let me know if you have any questions about the above. Thanks!

kevjumba replied:
Hey @seanlip,
Could you clarify what this means: " You are trying to swap out the one (or at most two) functions in email_services in tests, because the email service in the platforms layer should be doing nothing but making an external call, and you don't want to make external calls in the tests."

I understand that I shouldn't be making a call to mailgun in tests but the function usage should be the same right. In tests, I should still be calling email_services.send_mail/email_services.send_bulk_mail except prior to that call, I call self.swap to swap it with my own mock function.

kevjumba replied:
@seanlip So our structure leaves send_email_to_recipients(which does the post to mailgun api) in mailgun_email_services and moves send_mail and send_bulk_mail to core/domain/email_services(). I need to mock these functions instead of the mailgun functions since most of the tests use those functions. That makes sense right? Because you mentioned that I should create stubs for functions in mailgun_email_services in point b above.

seanlip replied:
Mock only send_email_to_recipients(). The only point of mocking is to avoid using the actual production APIs. As much of the code that's being tested should stay the same between dev/test/prod as possible, otherwise you end up testing something that has little relation to what runs in dev/prod.

kevjumba replied:
Hm, so currently all of the functions in testing uses send_mail and send_bulk_mail to send out the emails, which I moved out of mailgun_email_services into core/domain/email_services. I think it makes sense to mock these instead since this is still part of the API, just one layer up from the platform layer.

seanlip replied:
I have already explained, in my previous comment, why that is not the way to go. You will probably need to redo all of this.

kevjumba replied:
I redid it and the mock now only mocks core.platform.emails send_email_to_recipients(*args) api.

kevjumba committed:
Fix typo

seanlip commented:
Yes, correct. The tests would call the same functions or callers of those
functions.

The crucial point here is to figure out the correct API boundary. E.g. if
sending regular and bulk mails use the same Mailgun function then you may
only need one function in mailgun_email_services.

>

kevjumba committed:
Add a loggin handler and a testing for logging

kevjumba committed:
Change comments and address linting

kevjumba committed:
Merge remote-tracking branch 'upstream/develop' into migration-to-mailgun

kevjumba committed:
[skip ci] Change the api and tests

kevjumba committed:
[skip ci] Fix linting

kevjumba committed:
Move functionality from mailgun and gae_email_services to core/domain/email_services and edit the tests

kevjumba committed:
Fix issues and make tests pass

kevjumba committed:
Migrate more over

kevjumba committed:
Migrate the rest of the services that use mail_stub

kevjumba committed:
Remove gae_email_services and all references

kevjumba committed:
Add parsing functionality back in

kevjumba committed:
Fix linting issues in test_utils

kevjumba committed:
[skip ci] Linting fixes

kevjumba committed:
Address review issues

kevjumba committed:
Merge remote-tracking branch 'upstream/develop' into migration-to-mailgun

kevjumba committed:
Reorder the setup structure

kevjumba committed:
Fix errors I found in Files Changed tab

kevjumba committed:
Fix linting

kevjumba commented:
@seanlip PTAL!

seanlip commented on core/controllers/classifier_test.py:
```diff
@@ -152,8 +153,14 @@ def mock_get_classifier_training_job_by_id(_):
-        with can_send_emails_ctx, can_send_feedback_email_ctx:
-            with fail_training_job:
+        with can_send_emails_ctx, can_send_feedback_email_ctx, (
+            self.swap(
```
I'm not sure if you saw my other comment, but the only thing you should be swapping in these tests is the single function in mailgun_email_services.

Ideally EmailTestBase would handle that for you, so that inheriting EmailTestBase is sufficient and you don't need to manually do the swap each time. I can think of two possible approaches to investigate -- one is to dig into what happens when a `def test_...()` function is executed and see if you can find a way to overwrite the test harness framework to run those functions within a self.swap() context manager. The other is to monkeypatch mailgun_email_services.send_email_to_recipients by setting it to some other value in setUp, and setting it back to its original value in tearDown.

kevjumba replied:
Done, this was really cool! I didn't realize but overriding the test harness made it much easier! Thanks for the suggestion.

seanlip commented on core/domain/email_services.py:
```diff
@@ -17,10 +17,19 @@
 
+import logging
+import re
+from textwrap import dedent # pylint: disable=import-only-modules
```
In this codebase, Python stdlibs are imported at the module layer. So please just do `import textwrap`.

Also, if you think you need to use a pylint pragma -- heed the warning and don't use it. In general, if you add a pylint-disable the reviewer will ask you to remove it. The lint checks in the codebase should be adhered to by default.

kevjumba replied:
Done

seanlip commented on core/domain/email_services.py:
```diff
@@ -69,3 +78,219 @@ def get_feedback_thread_reply_info_by_user_and_thread_ids(user_id, thread_id):
+        reply_to_id: str. The unique id of the sender.
+
+    Raises:
+        Exception: if the configuration in feconf.py forbids emails from being
```
Exception: The configuration ...

(Phrase others similarly. Also there are other exceptions which do not seem to be recorded here.)

kevjumba replied:
Done

seanlip commented on core/domain/email_services.py:
```diff
@@ -69,3 +78,219 @@ def get_feedback_thread_reply_info_by_user_and_thread_ids(user_id, thread_id):
+        Exception: if mailgun api key is not stored in feconf.MAILGUN_API_KEY.
+        Exception: if mailgun domain name is not stored in
+            feconf.MAILGUN_DOMAIN_NAME.
+            (and possibly other exceptions, due to mail.send_mail() failures)
```
What is this referring to (after the refactor)?

Treat docstrings like code: read them carefully and make sure they are correct.

kevjumba replied:
Done

seanlip commented on core/domain/email_services.py:
```diff
@@ -69,3 +78,219 @@ def get_feedback_thread_reply_info_by_user_and_thread_ids(user_id, thread_id):
+            html_body.encode(encoding='utf-8'), bcc=[bcc], reply_to=reply_to)
+    else:
+        msg_title = 'MailgunService.SendMail'
+        # pylint: disable=division-operator-used
```
As usual -- do what you can do in order to drop the pragma.

I don't understand the error message, too. What's the problem here? If the problem is that pylint is not recognizing the triple-quotes correctly, an alternative might be to have a list of single-quoted strings and do a .join('\n') at the end.

kevjumba replied:
I don't really understand why but the text 'text/plain' uses a division operator which triggers the pylint error. This might be a bug in the pylint. I will file an issue.

kevjumba replied:
Huh actually I just ran it again and it's no longer an issue, wonder why it triggered before.

seanlip commented on core/domain/email_services.py:
```diff
@@ -69,3 +78,219 @@ def get_feedback_thread_reply_info_by_user_and_thread_ids(user_id, thread_id):
+        msg = msg_title + dedent(msg_body)
+        logging.info(msg)
+        logging.info(
+            'You are not currently sending out real email since this is a' +
```
email --> emails

kevjumba replied:
Done

seanlip commented on core/domain/email_services.py:
```diff
@@ -69,3 +78,219 @@ def get_feedback_thread_reply_info_by_user_and_thread_ids(user_id, thread_id):
+
+def send_bulk_mail(
+        sender_email, recipient_emails, subject, plaintext_body, html_body):
+    """Sends an email using mailgun api.
```
This is wrong. This file should be agnostic of platform.

kevjumba replied:
Done

seanlip commented on core/domain/email_services.py:
```diff
@@ -69,3 +78,219 @@ def get_feedback_thread_reply_info_by_user_and_thread_ids(user_id, thread_id):
+            plaintext_body.encode(encoding='utf-8'),
+            html_body.encode(encoding='utf-8'), bcc=[bcc], reply_to=reply_to)
+    else:
+        msg_title = 'MailgunService.SendMail'
```
Please drop all references to mailgun in this file. This is supposed to be a generic file which can work with any platform-provided email service.

kevjumba replied:
Done

seanlip commented on core/domain/email_services.py:
```diff
@@ -69,3 +78,219 @@ def get_feedback_thread_reply_info_by_user_and_thread_ids(user_id, thread_id):
+        raise ValueError(
+            'Malformed sender email address: %s' % sender_email)
+
+    # To send bulk emails we pass list of recipients in 'to' paarameter of
```
This stuff, which is platform-dependent and is likely to vary between different platforms, should be handled at the platform layer.

kevjumba replied:
Which part are you referring to? Are you talking about the exceptions that are raised? If so, do you think all of them should be in the platform layer?

kevjumba replied:
Done Migrated over

seanlip commented on core/domain/email_services.py:
```diff
@@ -69,3 +78,219 @@ def get_feedback_thread_reply_info_by_user_and_thread_ids(user_id, thread_id):
+        get_incoming_email_address(reply_to_id)
+        if reply_to_id else '')
+
+    if not constants.DEV_MODE:
```
All the "if dev mode" stuff should be dropped from this file and moved to the platform layer, or at least the part of the platform layer that decides what service to use to actually send the email.

In fact you could use this as a way to ensure that your implementation here is generic: consider the dev mode implementation as an alternative implementation to mailgun and have a separate module (as a sibling to mailgun) for it. Then in the platform-chooser file (core/platform/models.py), use the latter if we're in dev mode and the former if we're in prod mode.

kevjumba replied:
Done

kevjumba committed:
Address PR refactor of classifier test, linting not addressed yet, nor backend coverage

kevjumba committed:
Change the API to make mailgun and dev mode apis the same

kevjumba committed:
Fix email_manager, editor, classifier

kevjumba committed:
fix email dashboard

kevjumba committed:
Fix the rest of the tests

kevjumba committed:
Add in uncommitted changes in email_services

kevjumba committed:
Fix docstrings and linting

kevjumba committed:
Add dev mode attributes

kevjumba committed:
Fix issues

kevjumba committed:
Merge remote-tracking branch 'upstream/develop' into migration-to-mailgun

kevjumba committed:
Add tests to test_utils

kevjumba committed:
Dev mode email services

kevjumba committed:
Fix issues

kevjumba committed:
Remove pylint disable

kevjumba committed:
Add tests to make coverage 100% and also handling for unsuccessful send email action

kevjumba committed:
Merge remote-tracking branch 'upstream/develop' into migration-to-mailgun

kevjumba committed:
Fix some lint

kevjumba committed:
Move the email sectioning out into mailgun

kevjumba committed:
Remove print

kevjumba committed:
Merge remote-tracking branch 'upstream/develop' into migration-to-mailgun

kevjumba committed:
Fix some linting issues

kevjumba committed:
Fix lint

kevjumba committed:
Fix issues found in FilesChanged perusal

kevjumba committed:
Fix 1 lint issue

kevjumba commented:
@seanlip PTAL, I haven't fully figured out the tests required to run pre-release but the PR should be clean!

seanlip commented on core/controllers/classifier_test.py:
```diff
@@ -168,10 +167,10 @@ def mock_get_classifier_training_job_by_id(_):
                 # Check that there are no sent emails to either
                 # email address before posting json.
-                messages = self.mail_stub.get_sent_messages(
+                messages = self.email_services_mock.mock_get_sent_messages(
```
This should call the regular methods. (email_services.get_sent_messages() etc.). EmailTestBase should take care of the swap at the platform layer so that calling the regular method does the right thing.

If you call a mock, then we end up with a "skew" problem again, where the code you are testing is not the same as the code that runs in production.

This needs to be fixed in many places elsewhere, too.

kevjumba replied:
@seanlip So originally my idea was to swap out this line too since it makes sense that the API is the same. However, there isn't an email_services.get_sent_messages() equivalent for email_services since there are 2 types of email_services currently, one of which sends it to mailgun and the other of which logs to the terminal. I don't see a reasonable way of adding a get_sent_messages that does something valid in the API besides being a dummy function. 

Also this is following the trend of other swapping mocks where they write a function to swap out the important details and instead add something to a dictionary. However the only difference is they don't have that function because they didn't write an extra test-class so the dictionary they add to is actually a part of the class itself so they can just use it. 
An example would be in jobs_test.py where they mock the logging function:
```
   def test_failing_jobs(self):
        observed_log_messages = []

        def _mock_logging_function(msg, *args):
            """Mocks logging.error()."""
            observed_log_messages.append(msg % args)
```

seanlip replied:
Ah, sorry, you're right here, my bad! Scrap the previous comment, it doesn't apply for this particular method (though see the other notes for perhaps not having this as a separate mock class).

kevjumba replied:
No problem! Thanks for clarifying!

seanlip commented on core/domain/email_services.py:
```diff
@@ -69,3 +75,165 @@ def get_feedback_thread_reply_info_by_user_and_thread_ids(user_id, thread_id):
+        email_address: str. Email address to check.
+
+    Returns:
+        bool. Whether specified email address is valid.
```
Write in complete English clauses. In this case, there's a missing "the" before "specified".

Check your other docstrings, too.

kevjumba replied:
Done

seanlip commented on core/domain/email_services_test.py:
```diff
@@ -57,3 +62,198 @@ def test_reply_info_email_objects_are_created_and_queried_correctly(self):
+            self.assertEqual(messages[0].body, 'body')
+            self.assertEqual(messages[0].html, 'html')
+
+
```
Leave only one newline between adjacent methods.

kevjumba replied:
Done

seanlip commented on core/platform/email/dev_mode_email_services.py:
```diff
@@ -0,0 +1,119 @@
+        sender_email, recipient_emails, subject,
+        plaintext_body, html_body, bcc=None, reply_to=None,
+        recipient_variables=None):
+    """Logs out sent email details to terminal console in DEV_MODE, to model
```
"Logs out" is not really what you mean in this context, I think. Perhaps just "Logs"?

kevjumba replied:
Done

seanlip commented on core/platform/email/dev_mode_email_services.py:
```diff
@@ -0,0 +1,119 @@
+                    {"bob@example.com": {"first":"Bob", "id":1},
+                     "alice@example.com": {"first":"Alice", "id":2}}
+                subject = 'Hey, %recipient.first%’
+            More info at:
```
More info about this format at:

kevjumba replied:
Done

seanlip commented on core/platform/email/dev_mode_email_services.py:
```diff
@@ -0,0 +1,119 @@
+                latest/user_manual.html#batch-sending
+
+    Returns:
+        bool. Whether the email is logged successfully.
```
... is "sent" successfully.

kevjumba replied:
Done

seanlip commented on core/platform/email/mailgun_email_services.py:
```diff
@@ -14,138 +14,116 @@
+        resp = python_utils.url_open(req)
+        # The function url_open returns a file_like object that can be queried
+        # for the status code of the url query. If it is not 200, the mail query
+        # failed so we return False(this function did not complete
```
add space before '('

kevjumba replied:
Done

seanlip commented on core/platform/email/mailgun_email_services.py:
```diff
@@ -14,138 +14,116 @@
+            'html': html_body
+        }
+
+        if len(email_list) == 1:
```
Here and below, do you actually need the branching?

If you do, could also simplify to `data['to'] = ... if len(email_list) == 1 else ...`, and ditto below.

kevjumba replied:
Done, yes the post function, although opening the same url, sends a different request for singular email vs sending multiple urls so since batch sending vs normal sending is different, I don't want to use batch sending for a single email.

seanlip commented on core/platform/email/mailgun_email_services.py:
```diff
@@ -14,138 +14,116 @@
+        if recipient_variables:
+            data['recipient_variables'] = recipient_variables
+        else:
+            data['recipient_variables'] = {}
```
Simplify to `data['recipient_variables'] = recipient_variables or {}` (check that this works)

Or, if you want to be less fancy, have 
```
if recipient_variables is None:
    recipient_variables = {}
```
at the top of the function.

kevjumba replied:
Done

seanlip commented on core/tests/test_utils.py:
```diff
@@ -161,6 +163,100 @@ def check_image_png_or_webp(image_string):
+        self.recipient_variables = recipient_variables
+
+
+class EmailServicesMock(python_utils.OBJECT):
```
Drop this and move all the stuff into private methods in GenericEmailTestBase.

kevjumba replied:
Done

seanlip commented on core/tests/test_utils.py:
```diff
@@ -2282,6 +2375,74 @@ def assert_same_list_elements(self, phrases, stdout):
+    @classmethod
+    def setUpClass(cls):
+        super(GenericEmailTestBase, cls).setUpClass()
+        cls.email_services_mock = EmailServicesMock()
```
Don't need this. The test methods should not need to call the mock directly. Individual test functions may call stuff like self._get_received_email_messages() or similar (see above note about moving the functions into here).

kevjumba replied:
Done

seanlip commented on core/tests/test_utils.py:
```diff
@@ -2282,6 +2375,74 @@ def assert_same_list_elements(self, phrases, stdout):
+EmailTestBase = GenericEmailTestBase
+
+
+class MockLoggingHandler(python_utils.OBJECT):
```
Use same approach as mentioned above.

Basically the issue with the current approach is that a LoggingHandler isn't a thing. When you declare a mock, it should be the same type as an existing thing so that you can swap it out cleanly. In this case it looks like you're swapping at the method level, so just make sure the methods are available in your subclass.

Additionally, since the logging thing isn't related to emails, that should be part of the specific test class rather than the general infrastructure here.

kevjumba replied:
Ah interesting, this makes a lot of sense. Especially since it's only used in 1 or 2 test classes. 
Removed.

kevjumba committed:
Address review comments on PR

kevjumba committed:
Fix lint

kevjumba committed:
Merge remote-tracking branch 'upstream/develop' into migration-to-mailgun

kevjumba committed:
Fix issues found in FilesChanged

kevjumba committed:
Fix lint

kevjumba commented:
@seanlip PTAL! Thanks.

seanlip commented on core/controllers/email_dashboard_test.py:
```diff
@@ -47,46 +47,49 @@ def setUp(self):
-        self.logout()
+        swap_api = self.swap(feconf, 'MAILGUN_API_KEY', 'key')
+        swap_domain = self.swap(feconf, 'MAILGUN_DOMAIN_NAME', 'domain')
+        with swap_api, swap_domain:
```
Have the part within the "with" be as small as possible. I'm pretty certain you don't need it for everything that's here.

kevjumba replied:
Done

seanlip commented on core/controllers/email_dashboard_test.py:
```diff
@@ -719,11 +725,15 @@ def test_that_correct_emails_are_sent_to_max_n_recipients(self):
+                to=self.USER_A_EMAIL)
+            messages_b = self._get_sent_email_messages(
+                to=self.USER_B_EMAIL)
+            self.assertTrue((len(messages_a) == 1) or (len(messages_b) == 1))
```
Maybe also check that the other one is 0?

You can verify sorted(len(messages_a), len(messages_b)) against [0, 1]

kevjumba replied:
Done

seanlip commented on core/domain/email_manager_test.py:
```diff
@@ -2327,8 +2379,28 @@ def test_that_correct_rejected_voiceover_application_email_is_sent(self):
                 feconf.EMAIL_INTENT_VOICEOVER_APPLICATION_UPDATES)
 
+    def test_can_send_emails_is_false_logs_error(self):
+        """When feconf.CAN_SEND_EMAILS is false, send_rejected_voiceover should
```
send_rejected_voiceover? (That doesn't make sense semantically.)

kevjumba replied:
Done

seanlip commented on core/domain/email_manager_test.py:
```diff
@@ -2327,8 +2379,28 @@ def test_that_correct_rejected_voiceover_application_email_is_sent(self):
+            expected_log_message = 'This app cannot send emails to users.'
+
+            self.assertTrue(
+                observed_log_messages[0].startswith(expected_log_message))
```
Assert the entirety of observed_log_messages, not just the first element. In general, prefer the stronger validation check.

Also, in that spirit, why use startswith? Better to check equality of the whole thing.

kevjumba replied:
Done, I copied a previous use of mock logging. I thought it was to make sure it doesn't fail for edge cases but since tests pass with full equality it makes sense to not do that.

seanlip commented on core/domain/email_services.py:
```diff
@@ -69,3 +75,165 @@ def get_feedback_thread_reply_info_by_user_and_thread_ids(user_id, thread_id):
+    Returns:
+        bool. Whether the sender_email is valid.
+    """
+    # Checks the case where sender_email is of the form 'Jane <EMAIL>'.
```
Why are we checking for the number of names? This doesn't match either the docstring or the previous version of the checks, and I don't know that there is any requirement for the sender's name to have at most two space-separated words in it.

Try not to change the functionality/behaviour from before -- this is why you write tests. (You should probably add some backend tests that would have caught this error.)

kevjumba replied:
Done, in actuality the original implementation didn't really check for the whether the email or name existed but since that didn't really make sense with the method name, I added this check. Fixed now.

seanlip commented on core/domain/email_services.py:
```diff
@@ -69,3 +75,165 @@ def get_feedback_thread_reply_info_by_user_and_thread_ids(user_id, thread_id):
+    Raises:
+        Exception: The configuration in feconf.py forbids emails from being
+            sent.
+        Exception: If any recipient email addresses are malformed.
```
Exception: Any recipient ...

Ditto below.

kevjumba replied:
Done

seanlip commented on core/domain/email_services.py:
```diff
@@ -69,3 +75,165 @@ def get_feedback_thread_reply_info_by_user_and_thread_ids(user_id, thread_id):
+    if not feconf.CAN_SEND_EMAILS:
+        raise Exception('This app cannot send emails to users.')
+
+    if not is_email_valid(recipient_email):
```
One note -- if you don't expect these helper functions to be called from outside this module, prefix them with _.

(If the only thing calling them is tests, you can still test them by testing this function with appropriate arguments.)

kevjumba replied:
Done

seanlip commented on core/domain/email_services.py:
```diff
@@ -69,3 +75,165 @@ def get_feedback_thread_reply_info_by_user_and_thread_ids(user_id, thread_id):
+    if not is_sender_email_valid(sender_email):
+        raise ValueError(
+            'Malformed sender email address: %s' % sender_email)
+    bcc = [feconf.ADMIN_EMAIL_ADDRESS] if (bcc_admin) else None
```
You don't need the parens around bcc_admin.

kevjumba replied:
Done

seanlip commented on core/domain/email_services.py:
```diff
@@ -69,3 +75,165 @@ def get_feedback_thread_reply_info_by_user_and_thread_ids(user_id, thread_id):
+    reply_to = (
+        get_incoming_email_address(reply_to_id)
+        if reply_to_id else '')
+    resp = platform_email_services.send_email_to_recipients(
```
Do not abbreviate. Use the full word for "resp" (is it "response"?).

kevjumba replied:
Done

seanlip commented on core/domain/email_services.py:
```diff
@@ -69,3 +75,165 @@ def get_feedback_thread_reply_info_by_user_and_thread_ids(user_id, thread_id):
+        html_body.encode(encoding='utf-8'), bcc, reply_to, None)
+    if not resp:
+        raise Exception(
+            ('Email to %s failed to send. Please check your email service ' +
```
Not sure what this means. What exactly do you want the recipient of this message to do? The message itself could probably be stated more clearly.

kevjumba replied:
Done, made it more clear.

seanlip commented on core/domain/email_services.py:
```diff
@@ -69,3 +75,165 @@ def get_feedback_thread_reply_info_by_user_and_thread_ids(user_id, thread_id):
+    Raises:
+        Exception: The configuration in feconf.py forbids emails from being
+            sent.
+        Exception: If any recipient email addresses are malformed.
```
Same thing re formatting of exception descriptions.

kevjumba replied:
Done

seanlip commented on core/domain/email_services.py:
```diff
@@ -69,3 +75,165 @@ def get_feedback_thread_reply_info_by_user_and_thread_ids(user_id, thread_id):
+        raise ValueError(
+            'Malformed sender email address: %s' % sender_email)
+
+    resp = platform_email_services.send_email_to_recipients(
```
Same thing re not abbreviating the variable name.

kevjumba replied:
Done

seanlip commented on core/domain/email_services_test.py:
```diff
@@ -57,3 +62,192 @@ def test_reply_info_email_objects_are_created_and_queried_correctly(self):
+        send_email_exception = (
+            self.assertRaisesRegexp(
+                Exception, 'This app cannot send emails to users.'))
+        mailgun_api = self.swap(feconf, 'MAILGUN_API_KEY', 'api')
```
Maybe name these swap_mailgun_api, swap_mailgun_domain, etc. throughout?

Semantically I'm not sure what "with mailgun_api" means.

kevjumba replied:
Done

seanlip commented on core/platform/email/dev_mode_email_services.py:
```diff
@@ -0,0 +1,119 @@
+        sender_email, recipient_emails, subject,
+        plaintext_body, html_body, bcc=None, reply_to=None,
+        recipient_variables=None):
+    """Logs sent email details to terminal console in DEV_MODE, to model
```
Prints information about sent emails to the terminal console, in order to model sending an email in development mode.

(Prioritize clarity of docstrings. Add words where needed. The issue with the current docstring is that it starts with "Logs sent" which suggests that we are sending logs somewhere.)

kevjumba replied:
Done

seanlip commented on core/platform/email/dev_mode_email_services_test.py:
```diff
@@ -0,0 +1,143 @@
+            ' environment.')
+
+        allow_emailing = self.swap(feconf, 'CAN_SEND_EMAILS', True)
+        swap_api = self.swap(feconf, 'MAILGUN_API_KEY', 'key')
```
Why do you need to swap this stuff in dev mode? Ditto elsewhere in this file.

kevjumba replied:
Thanks for the pointer, I just realized we talked about the API more and removed the mailgun exceptions from the other implementations. This was actually a problem in some other parts of the pr too so I solved all of those too!

seanlip commented on core/platform/email/mailgun_email_services.py:
```diff
@@ -14,138 +14,108 @@
+                latest/user_manual.html#batch-sending
+
+    Raises:
+        Exception: If the mailgun api key is not stored in
```
See other comments re format of exceptions.

kevjumba replied:
Done

seanlip commented on core/platform/email/mailgun_email_services.py:
```diff
@@ -14,138 +14,108 @@
     Returns:
-        Response from the server. The object is a file-like object.
-        https://docs.python.org/2/library/urllib2.html
+        bool. Whether the email is "sent" successfully, contingent on the
```
You don't need the inverted quotes around "sent" here. Mailgun does actually send the email.

What do you mean by the "contingent on the ..." part? If you're saying that True means "200", then you'll want to say "... successfully. (This is implied by the mailgun API returning a status code of 200.)" instead.

kevjumba replied:
Done

seanlip commented on core/platform/models.py:
```diff
@@ -216,20 +217,21 @@ def import_app_identity_services(cls):
+        if constants.DEV_MODE:
+            from core.platform.email import dev_mode_email_services
+            return dev_mode_email_services
+        if (feconf.EMAIL_SERVICE_PROVIDER ==
```
elif

kevjumba replied:
Done

seanlip commented on core/tests/test_utils.py:
```diff
@@ -2282,6 +2280,118 @@ def assert_same_list_elements(self, phrases, stdout):
+class GenericEmailTestBase(GenericTestBase):
+    """Base class for tests requiring email services."""
+
+    class EmailMessageMock(python_utils.OBJECT):
```
It is fine for this to be declared as a top-level class.

kevjumba replied:
Done

seanlip commented on core/tests/test_utils.py:
```diff
@@ -2282,6 +2280,118 @@ def assert_same_list_elements(self, phrases, stdout):
+        def __init__(
+                self, sender_email, recipient_email, subject, plaintext_body,
+                html_body, bcc=None, reply_to=None, recipient_variables=None):
+            self.sender = sender_email
```
Write either a docstring or a one-sentence comment above each attribute to explain fully what it is.

Probably a docstring is better since you can include types as well.

kevjumba replied:
Done

seanlip commented on core/tests/test_utils.py:
```diff
@@ -2282,6 +2280,118 @@ def assert_same_list_elements(self, phrases, stdout):
+        a mailgun api key, mailgun domain name and mocked version of
+        send_email_to_recipients().
+        """
+        with self.swap(feconf, 'MAILGUN_API_KEY', 'key'), (
```
If we're doing this automatically then why are these swaps also present in some of the other test files?

kevjumba replied:
Done, yep realized that and fixed it.

seanlip commented on core/tests/test_utils.py:
```diff
@@ -2282,6 +2280,118 @@ def assert_same_list_elements(self, phrases, stdout):
+            plaintext_body: str. The plaintext body of the email. Must be utf-8
+            html_body: str. The HTML body of the email. Must fit in a datastore
+                entity. Must be utf-8.
+            bcc: list(str)|None. List of bcc emails. Must be utf-8.
```
Here and elsewhere, describe the None case too.

kevjumba replied:
Done

seanlip commented on core/tests/test_utils.py:
```diff
@@ -2282,6 +2280,118 @@ def assert_same_list_elements(self, phrases, stdout):
+            recipient_variables=(
+                recipient_variables if (recipient_variables) else None))
+        for recipient_email in recipient_emails:
+            if recipient_email not in self.emails_dict:
```
You can make self.emails_dict a collection.defaultdict, and then you won't need this sort of set-up.

kevjumba replied:
Done

seanlip commented on core/tests/test_utils.py:
```diff
@@ -2282,6 +2280,118 @@ def assert_same_list_elements(self, phrases, stdout):
+        return self.emails_dict[to] if to in self.emails_dict else []
+
+    def _get_all_sent_email_messages(self):
+        """Gets the entire messages dictionary."""
```
Missing returns docstring. (Make sure it's clear about the entire type specification of what's being returned.)

kevjumba replied:
Done

seanlip commented on core/tests/test_utils_test.py:
```diff
@@ -425,6 +425,47 @@ def mock_getcwd():
+        messages = self._get_sent_email_messages(
+            to='b@b.com')
+        all_messages = self._get_all_sent_email_messages()
+        self.assertEqual(len(messages), len(all_messages))
```
Here and elsewhere, always prefer exact comparisons for the expected value. Somewhere, you should have a check that these are both equal to 1.

kevjumba replied:
Done

seanlip reviewed (changes requested):
Thanks @kevjumba -- took a thorough pass, remaining comments are small ones. PTAL, thanks!

seanlip commented:
Also @prasanna08 @nithusha21 @DubeySandeep @aks681 @Hudda PTAL as codeowners!

kevjumba committed:
Address review comments

kevjumba committed:
Merge remote-tracking branch 'upstream/develop' into migration-to-mailgun

kevjumba committed:
Fix lint issues

kevjumba committed:
Add last clarification in docstring

kevjumba commented:
@seanlip  Thanks for the review! I fixed all of the errors. PTAL!

seanlip commented on core/domain/email_services.py:
```diff
@@ -120,17 +120,12 @@ def is_sender_email_valid(sender_email):
-            re.search(sender_email_with_full_name_regex, sender_email))
+    # Regex for 'SENDER_NAME <SENDER_EMAIL_ADDRESS>' or 'email_address'.
+    sender_email_regex = (
+        r'^[a-zA-Z._][a-zA-Z._ ]* <.+@(\[?)[a-zA-Z0-9-.]+.([a-zA-Z]{2,3}|' +
```
I'm not sure I quite follow what this regex is checking. How was it derived and does this match the checks that were there previously?

> Done, in actuality the original implementation didn't really check for the whether the email or name existed but since that didn't really make sense with the method name, I added this check. Fixed now.

I'm not sure what you mean by "existed". Could you elaborate?

kevjumba replied:
Originally for mailgun nothing was checked. This sender_email_valid function did not exist. The only checks that existed for gae_email_services like this
```
    if not _is_email_valid(sender_email):
        return False

    split_sender_email = sender_email.split(' ')
    if len(split_sender_email) < 2:
        return False

    email_address = split_sender_email[-1]
    if not email_address.startswith('<') or not email_address.endswith('>'):
        return False
```
which checked the format but not if the actual email address was valid. The is_email_valid check didn't check for email format either but only checked if the email was a python basestring
```
    if not isinstance(email_address, python_utils.BASESTRING):
        return False

    stripped_address = email_address.strip()
    if not stripped_address:
        return False
```

Since firstly, mailgun allows formats other than the format for _is_sender_email_valid, I decided to combine the two in one regex statement. I also added regex checking to checking the actual email address. 
The regex is basically checking for characters before @and then a site after the @ sign. I tested the regex on a regex checker online.

seanlip replied:
I still don't understand the last part of the regex, though. What does the [0-9]{1,3} represent?

kevjumba replied:
So originally, this was supposed to be a way to guarantee there is a valid (site) after the .,
aka .com or .edu or .ch or something. However, now I think it might be too restrictive so I will just make sure there are characters after the final dot.

seanlip commented on core/tests/test_utils.py:
```diff
@@ -2280,37 +2281,66 @@ def assert_same_list_elements(self, phrases, stdout):
+            plaintext_body: str. The plaintext body of the email. Must be utf-8
+            html_body: str. The HTML body of the email. Must fit in a datastore
+                entity. Must be utf-8.
+            bcc: list(str)|str|None. Optional argument. List of bcc emails,
```
Do not allow polymorphic types please. Either list(str) or str, but not both.

Ditto elsewhere.

kevjumba replied:
Done

seanlip commented on core/tests/test_utils.py:
```diff
@@ -2379,13 +2407,18 @@ def _get_sent_email_messages(self, to):
+        """Gets the entire messages dictionary.
+
+        Returns:
+            dict(list(EmailMessageMock)). The dictionary containing all lists of
```
This is not how we write docstrings for dicts (and the typeinfo is not correct either). Please look at other examples in the codebase and make this description clearer.

kevjumba replied:
Fixed.

seanlip commented on core/tests/test_utils_test.py:
```diff
@@ -428,13 +428,6 @@ def mock_getcwd():
 class EmailMockTests(test_utils.EmailTestBase):
     """Class for testing EmailTestBase."""
 
-    def test_override_run_swaps_contexts(self):
```
Any reason for dropping this?

kevjumba replied:
Couldn't really figure out a clean way of checking for function swap change so I thought it would suffice to use the second function(it wouldn't work if the function swap didn't occur). 

However, with further thought, I took some inspiration from the way self.swap was implemented and made a check for the function swap.

DubeySandeep reviewed (approved):
LGTM for the codeowner files `scripts/linters/pylint_extensions_test.py`.

kevjumba committed:
Address review comments

kevjumba committed:
Add a test_util test back in

kevjumba commented:
@seanlip Fixed the issues, could you take a look at my reasoning for changing the way email syntax is checked? If you would like it is easy to revert to what it was before since before, there were no checks on mailgun so we can just remove the email checking.

kevjumba committed:
Fix bug in email dashboard, the payload doesn't exist

kevjumba committed:
Remove unused import

seanlip commented on core/tests/test_utils_test.py:
```diff
@@ -428,6 +429,15 @@ def mock_getcwd():
+        email_services.send_email_to_recipients() is correctly swapped to its
+        mock version when the testbase extends EmailTestBase.
+        """
+        currrent_function = getattr(email_services, 'send_email_to_recipients')
```
"current" is misspelled

Perhaps called_function or referenced_function is a better name though.

kevjumba replied:
Done

kevjumba committed:
Make the regex simpler

kevjumba committed:
Name change of variable

kevjumba committed:
Fix bugs, not sure how this ever worked before

kevjumba committed:
Fix some files changed errors

kevjumba commented:
@seanlip PTAL! it turns out the API for backend tests requires it to use self.payload in email_dashboard.py(post_json uses specially creates a payload object) which makes me wonder how this code ever worked but I think I managed to fix it.

seanlip commented on core/domain/email_services.py:
```diff
@@ -106,7 +106,9 @@ def _is_email_valid(email_address):
         return False
     # Regex for a valid email.
-    regex = r'^.+@(\[?)[a-zA-Z0-9-.]+.([a-zA-Z]{2,3}|[0-9]{1,3})(]?)$'
+    # Matches any characters before the at sign, a series of characters until a
```
Not grammatically clear.

Matches any characters before the "@" sign, a series of characters until a ".", and then a series of characters after the period.

kevjumba replied:
Done

seanlip commented on core/domain/email_services.py:
```diff
@@ -121,9 +123,10 @@ def _is_sender_email_valid(sender_email):
     sender_email_regex = (
-        r'^[a-zA-Z._][a-zA-Z._ ]* <.+@(\[?)[a-zA-Z0-9-.]+.([a-zA-Z]{2,3}|' +
-        r'[0-9]{1,3})(]?)>$')
+        r'^[a-zA-Z._][a-zA-Z._ ]* <.+@[a-zA-Z0-9-.]+\.([a-zA-Z]+|[0-9]+)>$')
```
Try to remove the duplication between this regex and the above one -- some parts are reused.

kevjumba replied:
Done

seanlip commented:
@prasanna08 @nithusha21 @aks681 PTAL as codeowners, thanks!

nithusha21 reviewed (approved):
LGTM for codeowner files!

kevjumba committed:
Revert "Fix bugs, not sure how this ever worked before"

This reverts commit ab49e31cc5767ac095aff777358d0c9fc519b74a.

kevjumba committed:
Revert "Fix bug in email dashboard, the payload doesn't exist"

This reverts commit a88bdeb14c2a3ef0631b412b7d75da40f1e653c2.

kevjumba committed:
Address review comments

kevjumba committed:
Fix some docstrings

kevjumba committed:
Fix lint

kevjumba committed:
Fix a spelling error

kevjumba commented:
@seanlip I ended up reverting the email dashboard bug changes because although the fix some of the bugs, as detailed in the design document/email it still doesn't make email dashboard "work". I addressed the review comments and fixed some more grammatical mistakes PTAL.

seanlip commented:
Looks good, thanks. I'll try to run a test tonight. You'll need to fix the backend coverage though.

kevjumba committed:
Make test coverage 100%

kevjumba committed:
Change the test because the test is bugging out on travis ci

kevjumba committed:
Merge remote-tracking branch 'upstream/develop' into migration-to-mailgun

seanlip commented on core/domain/email_services_test.py:
```diff
@@ -197,7 +197,7 @@ def test_email_not_sent_if_email_addresses_are_malformed(self):
 
         # Case when sender is malformed when calling send_bulk_mail.
-        malformed_sender_email = 'email sender_email'
+        malformed_sender_email = 'name email@email.com'
```
Can you explain this change? What was the problem before? Does it work locally?

seanlip commented:
@prasanna08 @aks681 could you please review this as codeowners?

prasanna08 reviewed (approved):
LGTM for CODEOWNER files.

seanlip reviewed (approved):
Tested and it looks like the integration works. Thanks @kevjumba!

(Please address the last open comment.)

aks681 reviewed (approved):
lgtm as codeowner
